### PR TITLE
remote: verify inlined ActionResult output contents against their digest

### PR DESCRIPTION
**Pull request by evilgensec**
### Problem

When a remote/HTTP cache returns an `ActionResult` with an output file's bytes inlined in `OutputFile.contents`, `parseActionResultMetadata` copies those bytes into `FileMetadata.contents` and `downloadOutputs` consumes them directly (for example `inMemoryOutputData.set(file.contents)` for in-memory outputs such as `unused_inputs_list`) without comparing them against `OutputFile.digest`.

Every other output channel is checked against its digest: the digest-keyed download path verifies content via `Utils.verifyBlobContents`, the only producer of `OutputDigestMismatchException`. The inline channel is the one path that isn't, so `--remote_verify_downloads` (default on) does not cover it. `StarlarkAction` sets `REMOTE_EXECUTION_INLINE_OUTPUTS` automatically for any action using `unused_inputs_list`, so this is on a common code path, and the same channel also feeds C++ include scanning and `jdeps`.

To be clear about scope: this is not a defense against a malicious action cache. The cache supplies the digest and the inline bytes together and has to be trusted regardless, since it can set `digest = compute(contents)` for any bytes it inlines. What this closes is a verification-consistency gap: a buggy or inconsistent cache or proxy that inlines stale or wrong bytes is trusted silently today instead of failing fast like every other output.

### Fix

In `parseActionResultMetadata`, for any output file with non-empty inline `contents`, compute the digest of those bytes and compare it to `OutputFile.digest`, throwing `OutputDigestMismatchException` on mismatch, the same guarantee `Utils.verifyBlobContents` gives the digest-keyed path. Verifying once at parse covers every inline consumer (in-memory outputs, `unused_inputs_list`, include scanning, `jdeps`). Output files without inline contents are unaffected.

### Tests

`RemoteExecutionServiceTest` covers matching inline contents (served from memory), mismatching contents (rejected), and the empty/non-empty boundary: non-empty contents paired with the empty-blob digest are rejected, while truly empty contents are not treated as inlined and still resolve via the digest.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 1]
```diff
@@ -1166,13 +1166,28 @@ static ActionResultMetadata parseActionResultMetadata(
     for (OutputFile outputFile : result.getOutputFilesList()) {
       Path localPath =
           remotePathResolver.outputPathToLocalPath(unicodeToInternal(outputFile.getPath()));
+      ByteString contents = outputFile.getContents();
+      if (!contents.isEmpty()) {
+        // Inlined output bytes are consumed directly (e.g. for in-memory outputs
+        // such as unused_inputs_list). They bypass the digest-keyed download path
+        // that verifies content via Utils.verifyBlobContents, so verify them here
+        // against the accompanying digest before they are trusted.
+        Digest contentDigest = digestUtil.compute(contents.toByteArray());
```
[the remark this answers is not in the document]
Comment (evilgensec) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java]: Done in 571a383 — switched to `digestUtil.compute(contents::writeTo)`, which streams the `ByteString` into the hashing `OutputStream` (the `compute(DeterministicWriter)` overload) and avoids the `toByteArray()` copy.

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 2]
```diff
@@ -1166,13 +1166,28 @@ static ActionResultMetadata parseActionResultMetadata(
...
+        if (!contentDigest.equals(outputFile.getDigest())) {
+          OutputDigestMismatchException e =
+              new OutputDigestMismatchException(outputFile.getDigest(), contentDigest);
+          e.setOutputPath(outputFile.getPath());
+          e.setLocalPath(localPath);
+          throw e;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (evilgensec) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1225]: Kept the `!contents.isEmpty()` check. `contents.size() == digest.getSizeBytes()` is bypassable: a tampered response can inline N bytes while advertising a different `digest.size_bytes`, so the sizes differ, verification is skipped, and the N unverified bytes are still consumed. Verifying whenever `contents` is non-empty covers every case where there is tamperable inline data. Empty inline contents have nothing to verify and, because proto3 `bytes` has no field presence, are indistinguishable from a non-inlined output whose contents are simply absent — verifying the empty case would reject every non-inlined output (`compute(empty)` would not match the real file digest). So non-empty is the correct and complete discriminator here.

**Review by fmeum — COMMENTED** [review]
Could you add test coverage that also includes the tricky empty/non-empty mismatch case?

I'm not speaking in any official capacity for this project, but while this is certainly a case of missing validation, I don't think that this constitutes a security issue: the AC is responsible for the inlining and by design has to be trusted anyway as it can inject arbitrary output files.

**Comment by evilgensec** [comment]
Thanks, that's a fair correction and I agree on the framing. A cache that can poison the inline contents can set `OutputFile.digest = compute(contents)` just as easily, so this doesn't defend against a malicious action cache, which has to be trusted either way. I've dropped the security framing from the description.

I'd still like to land it as a consistency fix. The inline channel is the only output path not checked against its digest, so a buggy or inconsistent cache or proxy that inlines stale or wrong bytes is trusted silently today instead of failing fast like every other output, and `--remote_verify_downloads` doesn't cover it. Verifying once at parse keeps the inline path consistent with the digest-keyed path for every consumer (in-memory outputs, `unused_inputs_list`, include scanning, `jdeps`).

Added test coverage in caff30fb, including the empty/non-empty cases you asked about:
- matching inline contents are accepted and served from memory
- mismatching contents throw `OutputDigestMismatchException`
- non-empty contents paired with the empty-blob digest throw, rather than being served as a 0-byte file via the `contents.isEmpty()` path
- truly empty contents are not treated as inlined and still resolve via the digest, so the check doesn't misfire

**Comment by evilgensec** [comment]
Friendly ping for a maintainer look. Coverage for the empty and non-empty mismatch cases is in (caff30fb) and all presubmits are green. As discussed with @fmeum I dropped the security framing; it stands on its own as a consistency fix so the inline output path is digest-checked like every other output path. Anything else needed before it can go in?

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java [hunk 3]
```diff
@@ -1187,13 +1166,28 @@ static ActionResultMetadata parseActionResultMetadata(
...
+      ByteString contents = outputFile.getContents();
+      if (!contents.isEmpty()) {
+        // Inlined output bytes are consumed directly (e.g. for in-memory outputs
+        // such as unused_inputs_list). They bypass the digest-keyed download path
+        // that verifies content via Utils.verifyBlobContents, so verify them here
+        // against the accompanying digest before they are trusted.
+        Digest contentDigest = digestUtil.compute(contents::writeTo);
+        if (!contentDigest.equals(outputFile.getDigest())) {
+          OutputDigestMismatchException e =
+              new OutputDigestMismatchException(outputFile.getDigest(), contentDigest);
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1220]: This should be gated by `--remote_verify_downloads` to be consistent with other blobs.

**Review by tjgq — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java (hunk 3 above)
Comment (evilgensec) [reply src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java:1220]: Good catch, gated in 5c82178c. `parseActionResultMetadata` and `getOrParseActionResultMetadata` now take a `verifyDownloads` flag and both call sites pass `remoteOptions.getRemoteVerifyDownloads()`, so the inline path honors `--remote_verify_downloads` like GrpcCacheClient and HttpCacheClient do. Flag defaults to true so default behavior is unchanged. Added a test covering the disabled case.

**Comment by evilgensec** [comment]
@tjgq the `--remote_verify_downloads` gating you asked for landed in 5c82178c, which is the commit the red presubmit actually ran on. `parseActionResultMetadata` and `getOrParseActionResultMetadata` both take a `verifyDownloads` flag, both call sites pass `remoteOptions.getRemoteVerifyDownloads()`, and `downloadOutputs_inlinedContents_mismatchingDigest_notVerifiedWhenDisabled` covers the disabled path.

I've rebased onto master. The branch was 641 commits behind and I believe both failures came from that rather than from this change:

- `passport-control-check-docs` runs `mint validate` over the branch's `docs/` tree against a `docs.json` fetched from bazel-contrib/bazel-docs at HEAD, and 14 commits had touched `docs/` between the merge base and that run. This PR changes no docs and adds no flag.
- `windows-windows-shard-5` was the only red shard of ten, with every Linux, macOS and RBE shard green. The added code is a digest comparison with no path or filesystem handling, so I don't see how it could be shard-specific to this diff.

Ready for another look.

**Comment by evilgensec** [comment]
Correcting my previous comment: the rebase did not clear the failures, it surfaced more. Presubkite #35790 is now red on all eight ubuntu-24.04 shards plus clang, rbe, both rocky-8 variants, ubuntu-22.04 and windows-shard-4. That breadth is deterministic, not the flaky single-shard pattern from before, so my claim that the failures came only from the branch being behind was wrong.

What I have checked so far: the rebased tree's diff against master is exactly this change and nothing else (`RemoteExecutionService.java` +34/−6, `RemoteExecutionServiceTest.java` +116/−0), so the cherry-pick did not silently revert anything upstream, and the call sites and `FileMetadata` arity are consistent.

My working theory is that the tests I added conflict with 4c929580 ("remote: trust in-memory outputs across incremental builds"), which landed after this branch's original merge base and changes exactly the in-memory-output behaviour `downloadOutputs_inlinedContents_*` asserts on. fd5c5ba3 also reworked the same region. I can't build Bazel locally to isolate which assertion breaks.

Treat this as needing work rather than as ready for review — I'll follow up once I've identified the specific conflict. Apologies for the noise.
